RFC: New "networkmanager.passthrough" structure (LP: #2080301)

daniloegea opened the pull request:
## Description

This PR proposes a new format for the `networkmanager.passthrough` section. The new format was a suggestion from the [LP#2080301](https://bugs.launchpad.net/netplan/+bug/2080301) bug reporter. It addresses the problem reported in the bug and is future proof against similar problems.

The main problem with the currently used format is that it's ambiguous. The keyfile `[group].key=value` format is encoded as a string separated by dots. As dots can be used in the key name, it becomes tricky to separate what is the group and what is the key name. A different character could be used but we could eventually hit the same problem.

Arguably we could do something like `[group-name-between-brackets]key-name: value`. Brackets are forbidden in group names (I think?!). But it still would be a change in the format so I opted for turning each entry into mappings.

Here is how it looks like:

```yaml
  nm-devices:
    NM-4bf69677-e326-49b4-8e37-2acf3074f6c7:
      renderer: NetworkManager
      networkmanager:
        uuid: "4bf69677-e326-49b4-8e37-2acf3074f6c7"
        name: "tuntap7"
        passthrough:
          connection:
            type: "tun"
            autoconnect: "false"
            interface-name: "tuntap7"
          proxy: {}
          tun: {}
          ipv4:
            method: "manual"
            address1: "4.3.2.1/24,1.2.3.254"
            dns: "1.2.3.254;4.3.2.254;"
          ipv6:
            addr-gen-mode: "default"
            method: "manual"
            address1: "1:2:3::4/64,1:2:3::abcd"
            dns: "1:2:3::abcd;abcd:3:2::1;"
```

With the integration of netplan and Network Manager on Ubuntu, tweaking the passthrough configuration might be necessary every now and then due to things missing in netplan. This format makes it easier to do that.

The old format is still accepted for backwards compatibility. The parser still accepts it but will emit the new format. So YAMLs generated by Network Manager will be updated if the user modifies the connection and it will still work with existing YAMLs if they are never updated.

**IMPORTANT**: the `datalist` data structure was replaced with hash tables. This will make the order in which elements are added to keyfiles unpredictable. I'm not 100% sure that ordering would cause problems with configuration from the `passthrough` section. My main motivation to drop datalists is that [GQuarks might leak memory](https://github.com/GNOME/glib/blob/main/glib/gquark.c#L295). We hit this issue in our config_fuzzer test due to the number of GQuarks needed when a big number of entries are added to the datalist.

A couple of Network Manager autopkgtests are failing due to the change in the structure and will need to be fixed as well.

I understand it's a big change and we might decide to not merge it so I made it an RFC.

## Checklist

- [ ] Runs `make check` successfully.
- [ ] Retains code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

daniloegea committed:
parse: new structure for "passthrough"

The networkmanager.passthrough section stores raw keyfile information
using a format that can be tricky to interpret. It will concatenate the
group and key names separated by dots. Unfortunately, Network Manager
accepts multiple dots in the middle of both the group and key names.
Because of that it's hard, if even possible, to determine where the
strings must be split.

Ex:
networkmanager:
  passthrough:
    group_name.key_name: value1
    group_name.key_name1: value2

The new format stores the same information in YAML mappings. By using
this format, parsing the group and key names is not needed anymore.

Ex:
networkmanager:
  passthrough:
    group_name:
      key_name1: value1
      key_name2: value2

The datalist data structure was replaced by hash tables. The entire
passthrough section is represented by a hash table indexed by the group
name. Each entry is also a hash table representing all the key/value
entries and it's indexed by the key name.

daniloegea committed:
tests: adapt tests to new passthrough format

daniloegea committed:
config_fuzzer: run generate -i against the entire dataset

Now that the datalist was replaced with hash tables, run the generator
with --ignore-errors against the entire dataset. The datalist was
leaking memory and causing ASAN to crash the test.

Also limit the value of path-cost to avoid triggering a g_assert(v <
G_MACUINT).

daniloegea committed:
tests: update config_fuzzer with new passthrough format

Now it will generate a mix of old and new formats in the passthrough
section.

daniloegea committed:
nm: fix crashes when errors are ignored

When --ignore-errors is used, some netdefs might arrive at the NM config
writers in a bad state. In such cases we just skip them.

Found with config_fuzzer.

slyon commented on tests/generator/test_passthrough.py:
```diff
@@ -58,6 +58,83 @@ def test_passthrough_basic(self):
 match-device=type:ethernet
+managed=1\n\n''')
+
+    def test_passthrough_basic_new_format(self):
```
nitpick: `_new_format` is very generic. Maybe better call it `test_passthrough_basic_mapping`.

(Also in other places where "new_format" is used.)

slyon commented on tests/generator/test_passthrough.py:
```diff
@@ -58,6 +58,83 @@ def test_passthrough_basic(self):
+            type: ethernet
+            permissions: ""
+          group2:
+            key: a
+            key: b''', skip_generated_yaml_validation=True)
```
question: should the result rather be `key=b` here. So that the value defined later wins?

also followup question: what happens if we mix and match the old format?

e.g.
```yaml
passthrough:
  connection:
    name: "another name"
  connection.name: "some name"
 ```

I'd argue the value defined in the new (mapping) format should take precendence. WDYT?

daniloegea replied:
hmm yeaaah it probably should be `key=b` in the end. But that's actually a problem in the `handle_generic_map` function. It will error out if it finds the same key with different values (see below). Arguably this logic should be used everywhere else then...

```
$ cat /tmp/fakeroot/etc/netplan/a.yaml
network:
  version: 2
  renderer: NetworkManager
  ethernets:
    renderer: NetworkManager
    enp5s0:
      dhcp4: true
      openvswitch:
        other-config:
          one: 1
          one: 2
$ netplan  get --root-dir /tmp/fakeroot
Command failed: /tmp/fakeroot/etc/netplan/a.yaml:10:11: Error in network definition: duplicate map entry 'one'
          one: 1
          ^
```

For the second question, mixed format is supported. Although, if the key is set using the old style first, the parser will ignore it if it's found again using the new format. The problem is the same described above, but for these cases I'm ignoring the error.

I could create a new `handle_generic_map` only for the passthrough settings... but in this case I suppose the same logic should be used across all the parser and not only here...

slyon commented on tests/generator/test_passthrough.py:
```diff
@@ -75,6 +152,61 @@ def test_passthrough_wifi(self):
+match-device=type:wifi
+managed=1\n\n''')
+
+    def test_passthrough_wifi_new_format(self):
```
nitpick: Maybe call it ` test_passthrough_wifi_mapping` to be more precise.

slyon commented on src/validation.c:
```diff
@@ -407,7 +407,14 @@ validate_netdef_grammar(const NetplanParser* npp, NetplanNetDefinition* nd, GErr
+            key = g_hash_table_lookup(group, "type");
+        }
+    }
+    if (nd->type == NETPLAN_DEF_TYPE_NM && (!nd->backend_settings.passthrough || !key))
```
I don't think we need the `!nd->backend_settings.passthrough` here any more. `key` is only defined if `nd->backend_settings.passthrough` was accessible in the previous if-statement.

slyon commented on tests/cli/test_get_set.py:
```diff
@@ -479,15 +479,17 @@ def test_set_delete_nm_passthrough(self):
+                permissions: ""
+              ipv4:
+                dns-search: ""''')
         ap_key = 'network.wifis.wlan0.access-points.SOME-SSID'
-        self._set([ap_key+'.networkmanager.passthrough.connection\\.permissions=null'])
+        self._set([ap_key+'.networkmanager.passthrough.connection.permissions=null'])
```
question: Do we need to change the escaping in existing `netplan get` calls, out in the wild? That might have the potential to break the snapd integration used in Ubuntu Core.

daniloegea replied:
Yes, that (the escaping with `\`) will not work anymore if you are using get and set with settings inside the `passthrough` section. We probably shouldn't backport this change to LTS releases (like, never), even if we don't give any guarantees about stability for `passthrough`  block.

slyon reviewed (commented):
Thank you very much for working on this. I think the new and improved mapping format makes sense and is more sustainable than the existing flat list of group+key-values, going forward.

As can be seen from the CI NetworkManager tests, we need to adopt some cases in NetworkManager, and we should do so in a compatible way, e.g. using "netplan get", which can handle the old and new format, so we won't run into issues when backporting this change to older series.

I like the way you kept it reading the old format! But in general, we should double-check that we don't break backwards compatibility. Especially around `netplan get/set` and corresponding DBus calls, as used by the snapd integration.

Overall, getting rid of the `GData` list data-structure is a win, to avoid memory leaks in upstream GLib. And having a mapping structure is much more readable than the current flat list. I don't think there are any ordering issues introduced by using hashmaps, as we have one hashmap per keyfile group. And also the `GKeyFIle` functions put things into the correct place (group) inside the resulting file automatically.

I left a bunch of comments inline, mostly minor stuff, but also some questions, which we need to think through.

daniloegea commented on src/parse-nm.c:
```diff
@@ -448,25 +448,29 @@ parse_bond_arp_ip_targets(GKeyFile* kf, GArray **targets_arr)
             if (klen == 0) {
                 /* empty group */
-                g_datalist_set_data_full(list, g_strconcat(groups[i], ".", NETPLAN_NM_EMPTY_GROUP, NULL), g_strdup(""), g_free);
+                g_hash_table_insert(*list, g_strdup(groups[i]), group);
```
In this particular case, `groups` is created by `g_key_file_get_groups` which will not add repetitions to the result.

Although, there is a different memory leak there: thanks to the `continue` in the next line it will not free the `keys` array.

daniloegea commented on src/parse.c:
```diff
@@ -548,22 +549,78 @@ handle_generic_datalist(NetplanParser *npp, yaml_node_t* node, const char* key_p
-        assert_type(npp, value, YAML_SCALAR_NODE);
-
         escaped_key = g_strescape(scalar(key), STRESCAPE_EXCEPTIONS);
         escaped_value = g_strescape(scalar(value), STRESCAPE_EXCEPTIONS);
```
oh, that's a good point. The `escaped_value` is only relevant when the value type is `scalar`. I'll move it to inside the `if` statement.

daniloegea commented on src/parse.c:
```diff
@@ -548,22 +549,78 @@ handle_generic_datalist(NetplanParser *npp, yaml_node_t* node, const char* key_p
+            if (!g_hash_table_contains(*groups, group)) {
+                GHashTable *keys = g_hash_table_new(g_str_hash, g_str_equal);
+                /* Only add to the hash map if it's not the empty group */
+                if (g_strcmp0(k, NETPLAN_NM_EMPTY_GROUP)) {
```
Well, we could do something like `if(k != NULL && k[0] != NETPLAN_NM_EMPTY_GROUP && k[1] == '\0')`... 

`strcmp` will return after checking the first byte in most cases, we would save a function call though (or maybe not even that if `strcmp` gets inlined) but I think we don't need to get into this level of optimization? :)

daniloegea commented on src/parse.c:
```diff
@@ -548,22 +549,78 @@ handle_generic_datalist(NetplanParser *npp, yaml_node_t* node, const char* key_p
+            }
+            handle_generic_map(npp, value, full_key, &group, NULL, error);
+            if (*error) {
+                g_debug("Passthrough handling error, ignoring...: %s", (*error)->message);
```
The error message will have plenty of details, I think it should be good enough:

```
** (process:54840): DEBUG: 11:24:28.321: Passthrough handling error, ignoring...: /tmp/fakeroot/etc/netplan/c.yaml:15:13: Error in network definition: duplicate map entry 'b'
            b: 1
            ^
```

daniloegea committed:
parse: fix uninitialized variable

Fixes the build on Fedora 41:

In file included from /usr/include/glib-2.0/glib.h:117,
                 from ../src/parse.c:25:
In function ‘g_autoptr_cleanup_generic_gfree’,
    inlined from ‘get_ip_family’ at ../src/parse.c:1918:22:
/usr/include/glib-2.0/glib/glib-autocleanups.h:32:3: error: ‘ip_str’ may be used uninitialized [-Werror=maybe-uninitialized]
   32 |   g_free (*pp);
      |   ^~~~~~~~~~~~
../src/parse.c: In function ‘get_ip_family’:
../src/parse.c:1918:22: note: ‘ip_str’ was declared here
 1918 |     g_autofree char *ip_str;
      |                      ^~~~~~
cc1: all warnings being treated as errors

daniloegea committed:
docs: update with new passthrough format

daniloegea commented:
Thank you, Lukas.

I addressed a bunch of your comments and left answers for many others.

I think the main highlights are:

- `netplan set` will not work for **updating** fields inside the passthrough block. The reason is the way `handle_generic_map` works when it finds the same fields with different values. It also doesn't work for `openvswitch.other-config` for example due to the same problem. Adding and nullifying fields works fine.
- I updated the documentation too

I changed a lot of stuff, it's a good idea to go over all the code again (sorry for that).

daniloegea committed:
ci/rpmbuild: install gdb

The RPM build started to fail on Fedora 41:

find-debuginfo: starting
Extracting debug info from 4 files
gdb-add-index: Failed to find a useable GDB binary

slyon commented:
FTR: This is an enhancement of https://github.com/canonical/netplan/pull/193
